src/alda/lisp/cram.clj

```diff
@@ -0,0 +1,41 @@
+(ns alda.lisp.cram)
+
+(in-ns 'alda.lisp)
+
+(def no-events
+  {:start {:offset (->AbsoluteOffset 0), :events []}})
+
+(defmacro resetting [vars & body]
+  (if (seq vars)
+    `(do ~@body)
+    (let [[x & xs] vars]
+      `(let [before# ~x]
+         (resetting ~xs ~@body)))))
+
+
+(defmacro calc-duration [& body]
+  `(resetting
+    [~'alda.lisp/*events*
+     ~'alda.lisp/*instruments*
+     ~'alda.lisp/*global-attributes*]
+    (let [data# (do (alter-var-root ~'#'alda.lisp/*events* (constantly ~alda.lisp/no-events))
+                    (alter-var-root ~'#'alda.lisp/*global-attributes* #(dissoc % :duration))
```

Comment:
I'm not sure this is necessary -- `*global-attributes*` is actually a map of offsets to attribute change events.

To illustrate how it works:

```
alda.lisp=> (score*)
{}
alda.lisp=> (tempo! 90)
#alda.lisp.GlobalAttribute{:offset #alda.lisp.AbsoluteOffset{:offset 0}, :attr :tempo, :val 90}
alda.lisp=> (part* "piano")
nil
alda.lisp=> (do (note (pitch :c)) (note (pitch :d)) (note (pitch :e)))
(#alda.lisp.Note{:offset #alda.lisp.AbsoluteOffset{:offset 1333.3333740234375}, :instrument "piano-m7CA6", :volume 1.0, :track-volume 0.7874015748031497, :panning 0.5, :midi-note 64, :pitch 329.6275569128699, :duration 600.0000183105469})
alda.lisp=> (volume! 80)
#alda.lisp.GlobalAttribute{:offset #alda.lisp.AbsoluteOffset{:offset 2000.0000610351562}, :attr :volume, :val 80}
alda.lisp=> *global-attributes*
{0 [[:tempo 90]], 2000.0000610351562 [[:volume 80]]}
```